src/ValueRaw.cs

```diff
@@ -0,0 +1,285 @@
+using System;
+using System.Runtime.InteropServices;
+
+namespace Wasmtime
+{
+    [StructLayout(LayoutKind.Explicit)]
+    internal unsafe struct ValueRaw
+    {
+        [FieldOffset(0)]
+        public int i32;
+
+        [FieldOffset(0)]
+        public long i64;
+
+        [FieldOffset(0)]
+        public float f32;
+
+        [FieldOffset(0)]
+        public double f64;
+
+        [FieldOffset(0)]
+        public fixed byte v128[16];
+
+        [FieldOffset(0)]
+        public nuint funcref;
+
+        [FieldOffset(0)]
+        public nuint externref;
+
+        public static IValueRawConverter<T> Converter<T>()
+        {
+            if (typeof(T) == typeof(int))
+            {
+                return (IValueRawConverter<T>)Int32ValueRawConverter.Instance;
+            }
+
+            if (typeof(T) == typeof(long))
+            {
+                return (IValueRawConverter<T>)Int64ValueRawConverter.Instance;
+            }
+
+            if (typeof(T) == typeof(float))
+            {
+                return (IValueRawConverter<T>)Float32ValueRawConverter.Instance;
+            }
+
+            if (typeof(T) == typeof(double))
+            {
+                return (IValueRawConverter<T>)Float64ValueRawConverter.Instance;
+            }
+
+            if (typeof(T) == typeof(Function))
+            {
+                return (IValueRawConverter<T>)FuncRefValueRawConverter.Instance;
+            }
+
+            if (typeof(T) == typeof(V128))
+            {
+                return (IValueRawConverter<T>)V128ValueRawConverter.Instance;
+            }
+
+            if (!typeof(T).IsValueType)
+            {
+                return GenericValueRawConverter<T>.Instance;
+            }
+
+            throw new InvalidOperationException($"Cannot convert type '{typeof(T).Name}' into a WASM parameter type");
+        }
+    }
+
+    internal interface IValueRawConverter<T>
+    {
+        public T? Unbox(IStore store, in ValueRaw valueRaw);
+
+        public void Box(IStore store, T value, ref ValueRaw valueRaw);
+    }
+
+    internal class Int32ValueRawConverter : IValueRawConverter<int>
+    {
+        public static readonly Int32ValueRawConverter Instance = new();
+
+        private Int32ValueRawConverter()
+        {
+        }
+
+        public int Unbox(IStore store, in ValueRaw valueRaw)
+        {
+            return valueRaw.i32;
+        }
+
+        public void Box(IStore store, int value, ref ValueRaw valueRaw)
+        {
+            valueRaw.i32 = value;
```

Comment:
It seems the corresponding Rust code of this structure (`ValRaw`) [always initializes at least the first 64 bit when setting a value](https://github.com/bytecodealliance/wasmtime/blob/7a6fbe08987a7cb43c7e2a78ba5482bec8ab3c2a/crates/runtime/src/vmcontext.rs#L982-L993):
```rs
impl ValRaw {
    /// Creates a WebAssembly `i32` value
    #[inline]
    pub fn i32(i: i32) -> ValRaw {
        // Note that this is intentionally not setting the `i32` field, instead
        // setting the `i64` field with a zero-extended version of `i`. For more
        // information on this see the comments on `Lower for Result` in the
        // `wasmtime` crate. Otherwise though all `ValRaw` constructors are
        // otherwise constrained to guarantee that the initial 64-bits are
        // always initialized.
        ValRaw::u64((i as u32).into())
    }
```

However, this is not mentioned in the C header file for `wasmtime_val_raw_t`.

Do you know whether this also applies for the values in the C API, i.e. whether we would also need to set the first 64-bit, rather than only the first 32 bits for 32-bit values (so that this case, the next 32 bits will be set to zero)?

Reply:
I updated the code to always initialize the first 64-bit, so that we match the behavior of the Rust code, as this change doesn't seem to have effects on the performance (at least on 64-bit systems).

Additionally I added a check to ensure we throw a `PlatformNotSupportedException` on Big Endian systems when retrieving a converter, since we would need to implement separate converters for that, which implement switching endianness (as the values in the structure are always in little endian, if I understand the code comments/documentation correctly).

Reply:
> as the values in the structure are always in little endian, if I understand the code comments/documentation correctly

It seems bytecodealliance/wasmtime#4035 mistakenly added the notes about the little-endian storage to `wasmtime_valunion` instead of `wasmtime_val_raw`; I've opened bytecodealliance/wasmtime#5303 for that.